### PR TITLE
Quest challenge rework

### PR DESCRIPTION
**Pull request by Lordfirespeed**

### CODE REVIEW

File: server/src/socket/index.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    try {
+      ;({ user, scope } = await TokenVault.getUserAndScopeClaims(decodedPayload))
+    } catch (error) {
+      return cb(false)
+    }
+
+    if (scope.length === 0) return cb(false)
+
+    this.connectedUser = user
+    this.userRoles = scope as UserRole[]
```
Comment (Lordfirespeed) [inline server/src/socket/index.ts:73]: this is a flagrant misuse of the `scope` parameter :sob: :joy:

File: client/src/app/volunteer/(volunteer)/volunteer-page.tsx [hunk 2]
```diff
@@ -56,25 +59,44 @@ export default function Volunteer() {
     });
```
[the remark this answers is not in the document]
Comment (JacobWrenn) [reply client/src/app/volunteer/(volunteer)/volunteer-page.tsx]: Removed, was used for QR downloads

File: client/src/components/leaderboard/mega-chart.tsx [hunk 3]
```diff
@@ -50,7 +52,7 @@ export default function MegaChart() {
   });
 
   const dataset = {
-    labels: megateams?.map((team: any) => team.megateam_name),
+    labels: megateams?.map((team: any) => team.megateam_name.split(" ")[0]),
```
[the remark this answers is not in the document]
Comment (JacobWrenn) [reply client/src/components/leaderboard/mega-chart.tsx:55]: Getting the chartjs to play nice rendering the point labels, megateam logos and bars all nicely at the right heights on desktop and mobile was horrendous last year. Truncating the team names is infinitely better than how it looked with them overflowing

[the remark this answers is not in the document]
Comment (Lordfirespeed) [reply client/src/components/leaderboard/mega-chart.tsx]: we should just abbreviate the names with a `Map`, like 
- CotE 
- Atlantis
- Moon
- MI

File: client/src/components/tabbed-page.tsx [hunk 4]
```diff
@@ -31,22 +33,20 @@ export default function TabbedPage({
...
   const path = usePathname();
-  const [open, setOpen] = useState(false);
-  const { mutateUser } = useUser();
+  const [open, setOpen] = React.useState(false);
+  const { mutateUser } = useMegateamsContext();
 
   async function signOut() {
     await fetchMegateamsApi("/auth/logout", { method: "POST" });
-    mutateUser(null);
+    await mutateUser(null);
```
[the remark this answers is not in the document]
Comment (Lordfirespeed) [reply client/src/components/tabbed-page.tsx:41]: more importantly: this doesn't work! we need to actually redirect

File: client/src/app/volunteer/(volunteer)/preset.tsx [hunk 5]
```diff
@@ -1,95 +1,122 @@
...
-          }),
-          headers: { "Content-Type": "application/json" },
-        }
+      const { data: qr } = await fetchGuildsApi(
+        "/qr_codes/challenges/" + encodeURIComponent(id),
+        { method: "POST" }
       );
       setError(null);
-      displayQR(qr.name, qr.redemption_url, qr.category);
+      displayQR(qr.qrCodeId);
```
Comment (Lordfirespeed) [inline client/src/app/volunteer/(volunteer)/preset.tsx:36]: when server sends new QR code over socket, show big picture of checkmark for a second or so, to allow volunteers to close QR window before an extra qr is displayed

File: server/src/common/check-for-quest-completion.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+export async function checkForQuestCompletion(user: User, challengeIds: number[]) {
+  const quests = await prisma.quest.findMany({
+    where: {
+      challenges: { some: { challengeId: { in: challengeIds } } },
+      usersCompleted: { none: { keycloakUserId: user.keycloakUserId } },
+    },
+    include: {
+      challenges: {
+        include: {
```
Comment (Lordfirespeed) [inline server/src/common/check-for-quest-completion.ts:11]:
```suggestion
        select: {
```

File: server/src/common/check-for-quest-completion.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+          qrCodes: {
+            include: {
```
Comment (Lordfirespeed) [inline server/src/common/check-for-quest-completion.ts:13]:
```suggestion
            select: {
```

**Review by JacobWrenn — APPROVED** [review]

**Review by ducomputingsociety — APPROVED** [review]

**Pull request merged**